internal/scaffold/role_test.go

```diff
@@ -82,6 +86,336 @@ func TestRoleCustomRules(t *testing.T) {
+	absRoleFile1 := RoleFilePath1 + "/deploy/role.yaml"
+	if err := ioutil.WriteFile(absRoleFile1, []byte(roleFile1), fileutil.DefaultFileMode); err != nil {
+		fmt.Printf("failed to update %v: %v", absRoleFile1, err)
+	}
+
+	RoleFilePath2 := "./testdata/testroles/valid_role2"
+	absRoleFile2 := RoleFilePath2 + "/deploy/role.yaml"
+
+	if err := ioutil.WriteFile(absRoleFile2, []byte(roleFile2), fileutil.DefaultFileMode); err != nil {
+		fmt.Printf("failed to update %v: %v", absRoleFile2, err)
+	}
+
+	testCases := []struct {
+		name           string
+		absProjectPath string
+		r              *Resource
+		roleScaffold   *Role
+		expError       error
+		existingRole   string
+		mergedRole     string
+	}{
+		{
+			name:           "Valid CustomRules-1",
+			absProjectPath: RoleFilePath1,
+			mergedRole:     RoleFilePath1 + "/mergedRole.yaml",
+			r: &Resource{
+				APIVersion: "charts.helm.k8s.io/v1alpha1",
+				Kind:       "Memcached",
+				FullGroup:  "charts.helm.k8s.io",
+				LowerKind:  "memcached",
+				Resource:   "memcacheds",
+			},
+			roleScaffold: &Role{
+				IsClusterScoped:  false,
+				SkipDefaultRules: true,
+				CustomRules: []rbacv1.PolicyRule{
+					{
+						APIGroups: []string{""},
+						Resources: []string{"namespaces"},
+						Verbs:     []string{"get"},
+					},
+					{
+						APIGroups: []string{""},
+						Resources: []string{"configmaps", "secrets"},
+						Verbs:     []string{"*"},
+					},
+					{
+						APIGroups: []string{""},
+						Resources: []string{"events"},
+						Verbs:     []string{"create"},
+					},
+					{
+						APIGroups: []string{""},
+						Resources: []string{"services"},
+						Verbs:     []string{"*"},
+					},
+					{
+						APIGroups: []string{"apps"},
+						Resources: []string{"statefulsets"},
+						Verbs:     []string{"*"},
+					},
+				},
+			},
+		},
+
+		{
+			name:           "Valid CustomRules-2",
+			absProjectPath: RoleFilePath2,
+			mergedRole:     RoleFilePath2 + "/mergedRole.yaml",
+			r: &Resource{
+				APIVersion: "cache.example.com/v1alpha1",
+				Kind:       "Mykind",
+				FullGroup:  "cache.example.com",
+				LowerKind:  "mykind",
+				Resource:   "mykinds",
+			},
+			roleScaffold: &Role{
+				IsClusterScoped:  false,
+				SkipDefaultRules: true,
+				CustomRules: []rbacv1.PolicyRule{
+					{
+						APIGroups: []string{""},
+						Resources: []string{"namespaces"},
+						Verbs:     []string{"get"},
+					},
+					{
+						APIGroups: []string{""},
+						Resources: []string{"configmaps", "secrets"},
+						Verbs:     []string{"*"},
+					},
+					{
+						APIGroups: []string{""},
+						Resources: []string{"events"},
+						Verbs:     []string{"create"},
+					},
+					{
+						APIGroups: []string{""},
+						Resources: []string{"serviceaccounts", "services"},
+						Verbs:     []string{"*"},
+					},
+					{
+						APIGroups: []string{"apps"},
+						Resources: []string{"deployments"},
+						Verbs:     []string{"*"},
+					},
+				},
+			},
+		},
```

Comment:
For both 1 and 2, the original role and the merged role are identical files. Can we change 2 to also include a new CustomRule that does not exist in the original file to make sure the merge works in that case (i.e. we should see all the original rules and the new rule in one file)?

Ideally, we would also have rules in the original file that use all of `apiGroups`, `verbs`, `resourceNames` and `nonResourceURLs`, and then have rules in CustomRules that are nearly identical, but where we change just one value in each of those lists to make sure we always add a new rule rather than appending to the existing one. It's hard to explain, so ping me if you need clarification.

Reply:
Agreed, Will introduce testcase with a new custom rule. And another testcase with `resourceNames` and `nonResourceURLs`